File: org.ai4fm.proofprocess.isabelle.core/src/org/ai4fm/proofprocess/isabelle/core/analyse/ProofAnalyzer.java
package org.ai4fm.proofprocess.isabelle.core.analyse;

import isabelle.Command;
import isabelle.Document;
import isabelle.Command.State;
import isabelle.Text.Range;
import isabelle.XML.Tree;
import isabelle.scala.DocumentRef;

import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import org.ai4fm.filehistory.FileVersion;
import org.ai4fm.proofprocess.Attempt;
import org.ai4fm.proofprocess.Intent;
import org.ai4fm.proofprocess.Loc;
import org.ai4fm.proofprocess.Proof;
import org.ai4fm.proofprocess.ProofElem;
import org.ai4fm.proofprocess.ProofEntry;
import org.ai4fm.proofprocess.ProofFeature;
import org.ai4fm.proofprocess.ProofInfo;
import org.ai4fm.proofprocess.ProofProcessFactory;
import org.ai4fm.proofprocess.ProofSeq;
import org.ai4fm.proofprocess.ProofStep;
import org.ai4fm.proofprocess.Term;
import org.ai4fm.proofprocess.Trace;
import org.ai4fm.proofprocess.core.analyse.ProofMatcher;
import org.ai4fm.proofprocess.core.store.IProofStore;
import org.ai4fm.proofprocess.isabelle.IsabelleCommand;
import org.ai4fm.proofprocess.isabelle.IsabelleProofProcessFactory;
import org.ai4fm.proofprocess.isabelle.IsabelleTrace;
import org.ai4fm.proofprocess.isabelle.core.IsabelleProofPlugin;
import org.ai4fm.proofprocess.isabelle.core.parse.CommandParser;
import org.ai4fm.proofprocess.isabelle.core.parse.SnapshotReader;
import org.ai4fm.proofprocess.isabelle.core.parse.TermParser;
import org.ai4fm.proofprocess.log.ProofLog;
import org.ai4fm.proofprocess.project.Project;
import org.ai4fm.proofprocess.project.core.ProofHistoryManager;
import org.ai4fm.proofprocess.project.core.ProofManager;
import org.ai4fm.proofprocess.project.core.store.ProjectProofStore;
import org.ai4fm.proofprocess.project.core.util.ProofProcessUtil;
import org.ai4fm.proofprocess.project.core.util.ResourceUtil;
import org.eclipse.core.resources.IProject;
import org.eclipse.core.runtime.Assert;
import org.eclipse.core.runtime.CoreException;
import org.eclipse.core.runtime.IPath;
import org.eclipse.core.runtime.IProgressMonitor;
import org.eclipse.core.runtime.IStatus;
import org.eclipse.core.runtime.Path;
import org.eclipse.core.runtime.Status;

import scala.collection.JavaConversions;

/**
 * 
 * @author Andrius Velykis
 */
public class ProofAnalyzer {

	public IStatus analyze(Set<Command> changedCommands, Document.State docState, 
			IProgressMonitor monitor) throws CoreException {
	
		SnapshotReader reader = new SnapshotReader(changedCommands, docState);
		List<List<State>> proofStates = reader.readProofStates();
		if (proofStates.isEmpty()) {
			// nothing found to analyse
			return Status.OK_STATUS;
		}
		
		for (List<State> proofState : proofStates) {
			Set<Command> proofCommands = filterProofCommands(proofState, changedCommands);
			
			// TODO what if analyzer returns CANCEL_STATUS?
			analyze(proofState, proofCommands, reader, monitor);
		}
		
		return Status.OK_STATUS;
	}
	
	private Set<Command> filterProofCommands(List<State> proofState, Set<Command> filter) {
		
		Set<Command> proofCmds = new HashSet<Command>();
		for (State cmdState : proofState) {
			proofCmds.add(cmdState.command());
		}
		
		// keep just the commands in the filter
		proofCmds.retainAll(filter);
		
		return proofCmds;
	}
	
	private IStatus analyze(List<State> proofState, Set<Command> changedCommands,
			SnapshotReader reader, IProgressMonitor monitor) throws CoreException {

		Assert.isLegal(!proofState.isEmpty());
		
		State lastProofStep = proofState.get(proofState.size() - 1);
		Command lastCommand = lastProofStep.command();
		
		Integer commandStart = reader.getCommandStart(lastCommand);
		String documentText = reader.getDocumentText(lastCommand);
		if (commandStart == null || documentText == null) {
			// invalid command, no longer in the snapshot, so just ignore the proof
			return Status.OK_STATUS;
		}
		
		DocumentRef docRef = new DocumentRef(lastCommand.node_name());
		
		String filePathStr = docRef.getNode();
		IPath filePath = Path.fromOSString(filePathStr);
		
		IProject project = ResourceUtil.findProject(filePath);
		if (project == null) {
			// cannot locate the project, therefore cannot access proof process model
			return IsabelleProofPlugin.error("Unable to locate project for resource " + filePath, null);
		}
		
		Project proofProject = ProofManager.getProofProject(project, monitor);
		
		int commandEnd = commandStart.intValue() + lastCommand.range().stop();
		FileVersion fileVersion = ProofHistoryManager.syncFileVersion(
				project, filePath, documentText, commandEnd, monitor);
		
		analyzeEntry(new ProjectProofStore(proofProject), proofState, fileVersion);
		
		// FIXME retrieve proof entries from the analysis
		Map<State, ProofEntry> proofEntries = new HashMap<State, ProofEntry>();
		logActivity(project, proofState, changedCommands, proofEntries, monitor);
		
		// FIXME
		try {
			proofProject.eResource().save(ProofManager.SAVE_OPTIONS);
		} catch (IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}

		return Status.OK_STATUS;
	}
	
	private void analyzeEntry(IProofStore proofStore, List<State> proofState, FileVersion fileVersion) {
		
		// Assume that the first step in any proof is the "declaration" command, e.g. "lemma ..."
		State initialGoalState = proofState.get(0);
		List<Term> initialGoals = parseGoals(initialGoalState);
		
		if (initialGoals.isEmpty()) {
			// no initial goals - no proof
			return;
		}
		
		// find the proof for this
		ProofMatcher proofMatcher = new ProofMatcher();
		// TODO extract proof label
		Proof proof = proofMatcher.findCreateProof(proofStore, null, initialGoals);
		
		// Continue analysing the proof commands.
		// Note that the first State is excluded, as it is the "proof declaration command"
		List<State> remainingState = proofState.subList(1, proofState.size());
		if (remainingState.isEmpty()) {
			// no proof steps afterwards
			return;
		}
		
		// Convert to Proof Process steps - they will be matched to the existing
		// proof process records afterwards.
		List<ProofEntry> ppState = createProofSteps(proofMatcher, proofStore, proof, 
				remainingState, fileVersion);
		
		// TODO export State-Entry matchings for Activities
		proofMatcher.findCreateProofTree(proofStore, proof, ppState);
		
		// TODO save the proofProject
	}
	
	private List<ProofEntry> createProofSteps(ProofMatcher proofMatcher, IProofStore proofStore, 
			Proof proof, List<State> proofState, FileVersion fileVersion) {
		
		List<ProofEntry> entries = new ArrayList<ProofEntry>();
		
		List<Term> stepInGoals = proof.getGoals();
		for (State cmdState : proofState) {
			ProofEntry entry = createProofStep(
					proofMatcher, proofStore, fileVersion, stepInGoals, cmdState);
			entries.add(entry);
			
			stepInGoals = entry.getProofStep().getOutGoals();
		}
		
		return entries;
	}
	
	private ProofEntry createProofStep(ProofMatcher proofMatcher, IProofStore proofStore,
			FileVersion fileVersion, List<Term> stepInGoals, State commandState) {
		
		ProofInfo info = ProofProcessFactory.eINSTANCE.createProofInfo();
		info.setNarrative("Tactic: " + commandState.command().name());
		
		Intent intent = proofStore.getIntent("Tactic Application");
		info.setIntent(intent);
		
		// TODO set features
		List<ProofFeature> inFeatures = info.getInFeatures();
		List<ProofFeature> outFeatures = info.getOutFeatures();
		
		ProofStep step = ProofProcessFactory.eINSTANCE.createProofStep();
		step.setTrace(createProofStepTrace(commandState));
		step.setSource(createProofStepLoc(fileVersion, commandState));
		
		List<Term> inGoals = step.getInGoals();
		// copy the goals defensively because inGoals is a containment ref
		inGoals.addAll(proofMatcher.copyTerms(stepInGoals));
		
		List<Term> outGoals = step.getOutGoals();
		outGoals.addAll(parseGoals(commandState));
		
		// create tactic application attempt
		ProofEntry entry = ProofProcessFactory.eINSTANCE.createProofEntry();
		entry.setInfo(info);
		entry.setProofStep(step);
		
		return entry;
	}
	
	private List<Term> parseGoals(State commandState) {
		
		List<Term> parsedTerms = new ArrayList<Term>();
		
		for (Tree result : JavaConversions.asJavaIterable(commandState.results().values())) {
			List<Term> parsed = TermParser.parseGoals(result);
			parsedTerms.addAll(parsed);
		}
		
		return parsedTerms;
	}
	
	private Trace createProofStepTrace(State commandState) {
		
		IsabelleTrace trace = IsabelleProofProcessFactory.eINSTANCE.createIsabelleTrace();
		
		// parse the command
		IsabelleCommand cmd = CommandParser.parse(commandState);
		trace.setCommand(cmd);
		
		// TODO add simplifier lemmas
		List<String> simpLemmas = trace.getSimpLemmas();

		return trace;
	}
	
	private Loc createProofStepLoc(FileVersion fileVersion, State commandState) {
		
		Range cmdRange = commandState.command().range();
		int length = cmdRange.stop() - cmdRange.start();
		return ProofProcessUtil.createTextLoc(fileVersion, cmdRange.start(), length);
	}
	
	private Attempt createAttempt(List<? extends ProofElem> proofState) {
		Attempt attempt = ProofProcessFactory.eINSTANCE.createAttempt();
		
		if (!proofState.isEmpty()) {
			
			ProofElem rootElem;
			if (proofState.size() > 1) {
				// wrap the proof state into a sequential group
				ProofSeq seq = ProofProcessFactory.eINSTANCE.createProofSeq();
				seq.getEntries().addAll(proofState);
				
				ProofInfo info = ProofProcessFactory.eINSTANCE.createProofInfo();
				seq.setInfo(info);
				
				rootElem = seq;
			} else {
				rootElem = proofState.get(0);
			}
			
			attempt.setProof(rootElem);
		}
		
		return attempt;
	}
	
	
	private void logActivity(IProject project, List<State> proofState,
			Set<Command> changedCommands, Map<State, ProofEntry> proofEntries,
			IProgressMonitor monitor) throws CoreException {

		ProofLog proofLog = ProofManager.getProofLog(project, monitor);
		ProofActivityLogger activityLogger = new ProofActivityLogger(proofLog);

		activityLogger.logProof(proofState, changedCommands, proofEntries);
	}
	
}

File: org.ai4fm.proofprocess.isabelle.core/src/org/ai4fm/proofprocess/isabelle/core/analysis/ProofAnalyzer.scala
package org.ai4fm.proofprocess.isabelle.core.analysis

import org.ai4fm.filehistory.FileHistoryProject
import org.ai4fm.proofprocess.Attempt
import org.ai4fm.proofprocess.Loc
import org.ai4fm.proofprocess.ProofEntry
import org.ai4fm.proofprocess.ProofStore
import org.ai4fm.proofprocess.Term
import org.ai4fm.proofprocess.core.analysis.ProofAttemptMatcher.findCreateAttempt
import org.ai4fm.proofprocess.core.analysis.ProofAttemptMatcher.findCreateProof
import org.ai4fm.proofprocess.core.analysis.ProofEntryMatcher
import org.ai4fm.proofprocess.core.util.PProcessUtil
import org.ai4fm.proofprocess.isabelle.core.IsabellePProcessCorePlugin.error
import org.ai4fm.proofprocess.isabelle.core.parse.CommandResults
import org.ai4fm.proofprocess.isabelle.core.parse.ParsedProof
import org.ai4fm.proofprocess.isabelle.core.parse.ProofEntryReader
import org.ai4fm.proofprocess.isabelle.core.parse.SnapshotReader
import org.ai4fm.proofprocess.isabelle.core.parse.SnapshotReader.ProofData
import org.ai4fm.proofprocess.project.core.PProcessDataManager
import org.ai4fm.proofprocess.project.core.PProcessDataManager.PProcessData
import org.ai4fm.proofprocess.project.core.ProofHistoryManager
import org.ai4fm.proofprocess.project.core.util.ProofProcessUtil
import org.ai4fm.proofprocess.project.core.util.ResourceUtil
import org.eclipse.core.resources.IProject
import org.eclipse.core.runtime.CoreException
import org.eclipse.core.runtime.IStatus
import org.eclipse.core.runtime.IProgressMonitor
import org.eclipse.core.runtime.Path
import org.eclipse.core.runtime.Status
import org.eclipse.emf.ecore.util.EcoreUtil

import isabelle.Command
import isabelle.Command.State
import isabelle.Document
import isabelle.eclipse.core.resource.URIThyLoad.toURINodeName


/**
  * @author Andrius Velykis 
  */
object ProofAnalyzer {

  @throws[CoreException]
  def analyze(docState: Document.State,
      changedCommands: Set[Command],
      ppData: IProject => PProcessData,
      monitor: IProgressMonitor): IStatus = {

    // TODO sort proofs?
    val (proofs, commandStarts) = SnapshotReader.readProofs(docState, changedCommands)

    proofs foreach processProof(changedCommands, ppData, commandStarts, monitor)

    Status.OK_STATUS;
  }

  /**
   * Processes the proof attempt by parsing proof entries and performing some analysis to
   * infer the ProofProcess Attempt and its structure.
   * 
   * If successful, the attempt and the activities are added to the ProofProcess data.
   */
  @throws[CoreException]
  private def processProof(changedCommands: Set[Command],
                           ppData: IProject => PProcessData,
                           commandStarts: Map[Command, Int],
                           monitor: IProgressMonitor)(proof: ProofData) =
    findProjectResource(proof.textData.name) match {

    // try to resolve file project
    case Some(project) => {

      val PProcessData(proofStore, proofLog, fileHistory) = ppData(project)

      // store file history and create a function to resolve command locations
      val textLocFun = syncFileVersion(project, fileHistory, commandStarts, proof, monitor)
      
      // parse the proof and try to extract proof entries and structure
      parsePPEntries(proofStore, proof.proofState, textLocFun) match {
        
        case Some((parsedProof, parseEntries)) => {
          
          // analyse the parsed proof and add to proof store if needed
          val (_, matchMapping) = analyzeProofAttempt(proofStore, parsedProof)
  
          // map snapshot entries to actually matched proof entries for logging
          val entryMatchMap = PProcessUtil.chainMaps(parseEntries, matchMapping)
          
          val proofState = proof.proofState.map(_.state)
          ProofActivityLogger.logProof(proofLog, entryMatchMap, changedCommands, proofState)

          // FIXME commit here or after all analysis? or somewhere else altogether?
          PProcessDataManager.commitTransaction(proofStore, monitor)
        }

        case None => // invalid proof - ignore PP analysis
      }
      
    }

    case _ => {
      // cannot locate proof project
      error(msg = Some("Unable to locate project for resource " + proof.textData.name.node))
    }
  }


  private def findProjectResource(document: Document.Node.Name): Option[IProject] = {

    val pathUri = document.uri
    
    if (pathUri.isAbsolute) {
      // URI encoding
      Option(ResourceUtil.findProject(pathUri))
    } else {
      // Path encoding
      val pathStr = document.node
      val filePath = Path.fromOSString(pathStr)
      Option(ResourceUtil.findProject(filePath))
    }
  }


  /**
   * Synchronises the current proof state file with file history.
   * 
   * Returns a function to compute `Loc` locations for commands within this file version.
   */
  private def syncFileVersion(project: IProject,
                              history: FileHistoryProject,
                              commandStarts: Map[Command, Int],
                              proof: ProofData,
                              monitor: IProgressMonitor): State => Loc = {

    val proofTextData = proof.textData
    val pathStr = proofTextData.name.node

    // sync file history version that corresponds to the given proof
    val historyMan = ProofHistoryManager.historyManager(project, history)
    val fileVersion = ProofHistoryManager.syncFileVersion(
      historyMan, history, project,
      pathStr, Some(proofTextData.documentText), Some(proofTextData.syncPoint), monitor)

    // location resolution function
    def textLoc(cmdState: State): Loc = {
      val cmd = cmdState.command
      // TODO ignore position at all instead of 0?
      val offset = commandStarts.getOrElse(cmd, 0)
      ProofProcessUtil.createTextLoc(fileVersion, offset, cmd.range.stop - cmd.range.start);
    }
    
    // return the location function
    textLoc
  }


  /**
   * Parses proof entries and structure from the captured raw proof state.
   */
  private def parsePPEntries(proofStore: ProofStore,
                             proofState: List[CommandResults],
                             cmdLoc: State => Loc) = {

    val entryReader = new ProofEntryReader {

      lazy val stepIntent = PProcessUtil.getIntent(proofStore, "Tactic Application")

      // TODO review
      override def cloneTerm(term: Term): Term = EcoreUtil.copy(term)
      
      // TODO ignore some properties, e.g. display?
      override def matchTerms(term1: Term, term2: Term): Boolean = EcoreUtil.equals(term1, term2);

      override def textLoc(cmdState: State) = cmdLoc(cmdState)
    }

    entryReader.readEntries(proofState)
  }
  
  private def analyzeProofAttempt(proofStore: ProofStore, parsedProof: ParsedProof)
      : (Attempt, Map[ProofEntry, ProofEntry]) = {
    
    val entryMatcher = ProofEntryMatcher()

    val targetProof =
      findCreateProof(entryMatcher)(proofStore, parsedProof.label, parsedProof.proofGoals)

    val attemptMapping =
      findCreateAttempt(entryMatcher)(parsedProof.proofGraph, targetProof)

    attemptMapping
  }

}
